cpp/src_prims/linalg/batched/matrix.h

```diff
@@ -95,80 +58,163 @@ __global__ void identity_matrix_kernel(T* I, int m) {
  *        id is the batch id.
  * 
  * @param[in]  in      Input vector
  * @param[out] out     Output vector
  * @param[in]  n_elem  Number of elements in the input vector
+ * @param[in]  period  Period of the difference
+ */
+template <typename T>
+__global__ void batched_diff_kernel(const T* in, T* out, int n_elem,
+                                    int period = 1) {
+  const T* batch_in = in + n_elem * blockIdx.x;
+  T* batch_out = out + (n_elem - period) * blockIdx.x;
+
+  for (int i = threadIdx.x; i < n_elem - period; i += blockDim.x) {
+    batch_out[i] = batch_in[i + period] - batch_in[i];
+  }
+}
+
+/**
+ * @brief Kernel to compute the second difference of batched vectors with given
+ *        periods (1 for simple difference, s for seasonal difference)
+ *
+ * @note: The thread id is the starting position in each vector and the block
+ *        id is the batch id.
+ * 
+ * @param[in]  in       Input vector
+ * @param[out] out      Output vector
+ * @param[in]  n_elem   Number of elements in the input vector
+ * @param[in]  period1  Period for the 1st difference
+ * @param[in]  period2  Period for the 2nd difference
  */
 template <typename T>
-__global__ void batched_diff_kernel(const T* in, T* out, int n_elem) {
+__global__ void batched_second_diff_kernel(const T* in, T* out, int n_elem,
+                                           int period1 = 1, int period2 = 1) {
   const T* batch_in = in + n_elem * blockIdx.x;
-  T* batch_out = out + (n_elem - 1) * blockIdx.x;
+  T* batch_out = out + (n_elem - period1 - period2) * blockIdx.x;
 
-  for (int i = threadIdx.x; i < n_elem - 1; i += blockDim.x) {
-    batch_out[i] = batch_in[i + 1] - batch_in[i];
+  for (int i = threadIdx.x; i < n_elem - period1 - period2; i += blockDim.x) {
+    batch_out[i] = batch_in[i + period1 + period2] - batch_in[i + period1] -
+                   batch_in[i + period2] + batch_in[i];
+  }
+}
+
+/**
+ * @brief Helper kernel for the allocation: computes the array of pointers to each
+ *        matrix in the strided batch
+ *
+ * @param[in]  A_dense     Raw data array
+ * @param[out] A_array     Array of strided pointers to A_dense
+ * @param[in]  batch_size  Number of matrices in the batch
+ * @param[in]  m           Number of rows of each matrix
+ * @param[in]  n           Number of columns of each matrix
+ */
+template <typename T>
+__global__ void fill_strided_pointers_kernel(T* A_dense, T** A_array,
+                                             int batch_size, int m, int n) {
+  int bid = blockIdx.x * blockDim.x + threadIdx.x;
+  if (bid < batch_size) {
+    A_array[bid] = A_dense + bid * m * n;
   }
 }
 
 /**
- * @brief The BatchedMatrix class provides storage and a number of linear
+ * @brief The Batched::Matrix class provides storage and a number of linear
  *        operations on collections (batches) of matrices of identical shape.
  */
 template <typename T>
-class BatchedMatrix {
+class Matrix {
```

Comment:
Because of the copy-ctor and assignment-operator, there could be temporary copies of `Matrix` objects getting created! Taking `b_solve` for eg: we'll have 4 objects created
1. temporary of `A` itself inside `inv`
2. `A.inv()`
3. `A.inv() * b`
4. `auto x = A.inv() * b`
And in case there are a lot of such operations, number of such temporary objects could be huge, thereby reducing our perf. That said, I'm ok defering this problem for later releases, if file an issue on this one.

Reply:
I created the copy constructor because we sometimes need copies, and the copy assignment operator because of the rule of three. I initially wrote a move constructor and assignment operator, but when switching to `device_buffer` I had to remove them because the buffers don't have move semantics...

@teju85 What do you think I should do:
- Add move semantics to `device_buffer` and then to `Batched::Matrix`, which should avoid the copies by moving the result of the functions to the variable?
- Remove the copy-ctor and assignment operator? I think it's impossible because the copy ctor of `device_buffer` is deleted, so it will delete the default copy-ctor and assignment operator of `Batched::Matrix` too and e.g `b_solve` won't compile.
- Change the functions to take a ref param instead of returning a value?

Is there something I'm not seeing? And is there any good reason why `device_buffer` doesn't have copy and move semantics?

Reply:
For now, I have filed issue #1692 for this. Let's discuss this there and do it the right way. Until then, let's leave this code as-is.